pkg/api/v1/reconcile/reconciler.go

```diff
@@ -54,43 +56,56 @@ func (r *reconciler) Reconcile(namespace string, desiredResources resources.Reso
 }
 
 func (r *reconciler) syncResource(ctx context.Context, desired resources.Resource, originalResources resources.ResourceList, transition TransitionResourcesFunc) error {
-	var overwriteExisting bool
+	var overwriteExisting, alreadyAttemptedUpdate bool
 	original := findResource(desired.GetMetadata().Namespace, desired.GetMetadata().Name, originalResources)
-	if original != nil {
-		// if this is an update,
-		// update resource version
-		// set status to 0, needs to be re-processed
-		overwriteExisting = true
-		resources.UpdateMetadata(desired, func(meta *core.Metadata) {
-			meta.ResourceVersion = original.GetMetadata().ResourceVersion
-		})
-		if desiredInput, ok := desired.(resources.InputResource); ok {
-			desiredInput.SetStatus(core.Status{})
-		}
-		// default transition policy: only perform an update if the Hash has changed
-		if transition == nil {
-			transition = func(original, desired resources.Resource) (b bool, e error) {
-				equal, ok := hashutils.HashableEqual(original, desired)
-				if ok {
-					return !equal, nil
-				}
 
-				// default behavior: perform the update if one if the objects are not hashable
-				return true, nil
-			}
-		}
-		if transition != nil {
-			needsUpdate, err := transition(original, desired)
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+
+		if alreadyAttemptedUpdate {
+			var err error
+			original, err = r.rc.Read(original.GetMetadata().Namespace, original.GetMetadata().Name, clients.ReadOpts{Ctx: ctx})
 			if err != nil {
 				return err
 			}
-			if !needsUpdate {
-				return nil
+		}
+
+		if original != nil {
+			// if this is an update,
+			// update resource version
+			// set status to 0, needs to be re-processed
+			overwriteExisting = true
+			resources.UpdateMetadata(desired, func(meta *core.Metadata) {
+				meta.ResourceVersion = original.GetMetadata().ResourceVersion
+			})
+			if desiredInput, ok := desired.(resources.InputResource); ok {
+				desiredInput.SetStatus(core.Status{})
+			}
+			// default transition policy: only perform an update if the Hash has changed
+			if transition == nil {
+				transition = func(original, desired resources.Resource) (b bool, e error) {
+					equal, ok := hashutils.HashableEqual(original, desired)
+					if ok {
+						return !equal, nil
+					}
+
+					// default behavior: perform the update if one if the objects are not hashable
+					return true, nil
+				}
+			}
+			if transition != nil {
+				needsUpdate, err := transition(original, desired)
+				if err != nil {
+					return err
+				}
+				if !needsUpdate {
+					return nil
+				}
 			}
 		}
-	}
-	_, err := r.rc.Write(desired, clients.WriteOpts{Ctx: ctx, OverwriteExisting: overwriteExisting})
-	return err
+		_, err := r.rc.Write(desired, clients.WriteOpts{Ctx: ctx, OverwriteExisting: overwriteExisting})
+		alreadyAttemptedUpdate = true
+		return err
+	})
```

Comment:
what do you think about using `retry.RetryOnConflict` or something similar (our own function?) similar to https://github.com/solo-io/autopilot/blob/da1d5cb0893789b305e3f38d118486a5f9bf5eed/pkg/ezkube/ensurer.go#L54